### PR TITLE
backport changes from #139 for current spec

### PR DESCRIPTION
**Pull request by walkah**
- backport better ucan header / payload types
- alter encode/decode & signing
- add `ucan.isValid`  and update test

### CODE REVIEW

File: src/ucan.ts [hunk 1]
```diff
@@ -176,12 +177,33 @@ export function decode(ucan: string): Ucan  {
...
+
+/**
+ * Encode the payload of a UCAN.
+ *
+ * @param payload The UcanPayload to encode
+ */
+export function encodePayload(payload: UcanPayload): string {
+  return base64.urlEncode(JSON.stringify({
+    ...payload,
+    prf: payload.prf.map(encode)
```
Comment (icidasset) [inline src/ucan.ts:205]: Does the server support a list at the moment? 👀

**Review by icidasset — APPROVED** [review]
Looks great, thanks! 👏 🎉 
Had one question regarding proofs.

File: src/ucan.ts (hunk 1 above)
Comment (walkah) [reply src/ucan.ts:205]: @icidasset oh... good catch! server *is* still single proof: https://github.com/fission-suite/fission/blob/main/fission-core/library/Fission/Web/Auth/Token/JWT/Types.hs#L132

thanks!

**Comment by walkah** [comment]
Huh. The fs concurrency test is failing ... I'll have a look at that too 🤔

File: src/ucan.ts [hunk 2]
```diff
@@ -1,35 +1,38 @@
 import { CryptoSystem } from 'keystore-idb/types'
 
+import * as did from './did'
 import * as keystore from './keystore'
 import { base64 } from './common'
 
-
 // TYPES
 
+export type Fact = any
```
Comment (walkah) [inline src/ucan.ts:9]: 🤔 this should get cleaned up to better match server's Fact type

**Comment by walkah** [comment]
@icidasset after doing a little more testing, I rolled back to single proof - keeps our types more explicit and in-step with server for now. we can update when server is ready. cool?

**Comment by icidasset** [comment]
Aight, sounds good 👍

**Pull request merged**